spark/common/src/main/scala/org/apache/sedona/stats/clustering/DBSCAN.scala

```diff
@@ -0,0 +1,153 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.sedona.stats.clustering
+
+import org.apache.sedona.stats.Util.getGeometryColumnName
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.sedona_sql.UDT.GeometryUDT
+import org.apache.spark.sql.sedona_sql.expressions.st_functions.{ST_Distance, ST_DistanceSpheroid}
+import org.apache.spark.sql.{Column, DataFrame, SparkSession}
+import org.graphframes.GraphFrame
+
+object DBSCAN {
+
+  private val ID_COLUMN = "__id"
+
+  /**
+   * Annotates a dataframe with a cluster label for each data record using the DBSCAN algorithm.
+   * The dataframe should contain at least one GeometryType column. Rows must be unique. If one
+   * geometry column is present it will be used automatically. If two are present, the one named
+   * 'geometry' will be used. If more than one are present and neither is named 'geometry', the
+   * column name must be provided. The new column will be named 'cluster'.
+   *
+   * @param dataframe
+   *   apache sedona idDataframe containing the point geometries
+   * @param epsilon
+   *   minimum distance parameter of DBSCAN algorithm
+   * @param min_pts
+   *   minimum number of points parameter of DBSCAN algorithm
+   * @param geometry
+   *   name of the geometry column
+   * @param includeOutliers
+   *   whether to include outliers in the output. Default is false
+   * @param useSpheroid
+   *   whether to use a cartesian or spheroidal distance calculation. Default is false
+   * @return
+   *   The input DataFrame with the cluster label added to each row. Outlier will have a cluster
+   *   value of -1 if included.
+   */
+  def dbscan(
+      dataframe: DataFrame,
+      epsilon: Double,
+      min_pts: Int,
+      geometry: String = null,
+      includeOutliers: Boolean = true,
+      useSpheroid: Boolean = false): DataFrame = {
+
+    // We want to disable broadcast joins because the broadcast reference were using too much driver memory
+    val spark = SparkSession.getActiveSession.get
```

Comment:
is here possible to have an empty session ?

Reply:
Only if the session crashed in some other test. The parent class initializes the spark session.